### PR TITLE
Use requests(-cache)

### LINKED ISSUE #11: Cache responses
Taking inspiration from [brok](https://github.com/smallhadroncollider/brok), I think we should have a cache that is persisted in CI, and only check links that haven't been checked in 24hrs (settable).  This will cut down on our DDOS-ing.

### PR DESCRIPTION
**Pull request by bollwyvl**
fixes #11

Part 1:
This adopts `requests` instead of relying on stdlib and six (though requests _also_ uses `six`, but no longer importing it). This isn't free, by any stretch, as it does bring along a number of transient deps, but Part 2 will be a lot less predictable without:

<details>
<pre>
# sorry for the conda noise... pip still doesn't have dry-run in 2020?
--- deps.old	2020-04-22 20:54:52.258914604 -0400
+++ deps.new	2020-04-22 20:54:40.494814215 -0400
@@ -11,7 +11,7 @@
     - nbconvert
     - nbformat
     - pytest
-    - six
+    - requests
 
 
 The following NEW packages will be INSTALLED:
@@ -20,13 +20,18 @@
   _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-0_gnu
   attrs              conda-forge/noarch::attrs-19.3.0-py_0
   bleach             conda-forge/noarch::bleach-3.1.4-pyh9f0ad1d_0
+  brotlipy           conda-forge/linux-64::brotlipy-0.7.0-py38h1e0a361_1000
   ca-certificates    conda-forge/linux-64::ca-certificates-2020.4.5.1-hecc5488_0
   certifi            conda-forge/linux-64::certifi-2020.4.5.1-py38h32f6830_0
+  cffi               conda-forge/linux-64::cffi-1.14.0-py38hd463f26_0
+  chardet            conda-forge/linux-64::chardet-3.0.4-py38h32f6830_1006
+  cryptography       conda-forge/linux-64::cryptography-2.8-py38h766eaa4_2
   decorator          conda-forge/noarch::decorator-4.4.2-py_0
   defusedxml         conda-forge/noarch::defusedxml-0.6.0-py_0
   docutils           conda-forge/linux-64::docutils-0.16-py38h32f6830_1
   entrypoints        conda-forge/linux-64::entrypoints-0.3-py38h32f6830_1001
   html5lib           conda-forge/noarch::html5lib-1.0.1-py_0
+  idna               conda-forge/noarch::idna-2.9-py_1
   importlib-metadata conda-forge/linux-64::importlib-metadata-1.6.0-py38h32f6830_0
   importlib_metadata conda-forge/noarch::importlib_metadata-1.6.0-0
   ipython_genutils   conda-forge/noarch::ipython_genutils-0.2.0-py_1
@@ -51,19 +56,24 @@
   pip                conda-forge/noarch::pip-20.0.2-py_2
   pluggy             conda-forge/linux-64::pluggy-0.13.1-py38_0
   py                 conda-forge/noarch::py-1.8.1-py_0
+  pycparser          conda-forge/noarch::pycparser-2.20-py_0
   pygments           conda-forge/noarch::pygments-2.6.1-py_0
+  pyopenssl          conda-forge/noarch::pyopenssl-19.1.0-py_1
   pyparsing          conda-forge/noarch::pyparsing-2.4.7-pyh9f0ad1d_0
   pyrsistent         conda-forge/linux-64::pyrsistent-0.16.0-py38h1e0a361_0
+  pysocks            conda-forge/linux-64::pysocks-1.7.1-py38h32f6830_1
   pytest             conda-forge/linux-64::pytest-5.4.1-py38h32f6830_0
   python             conda-forge/linux-64::python-3.8.2-he5300dc_6_cpython
   python_abi         conda-forge/linux-64::python_abi-3.8-1_cp38
   readline           conda-forge/linux-64::readline-8.0-hf8c457e_0
+  requests           conda-forge/noarch::requests-2.23.0-pyh8c360ce_2
   setuptools         conda-forge/linux-64::setuptools-46.1.3-py38h32f6830_0
   six                conda-forge/noarch::six-1.14.0-py_1
   sqlite             conda-forge/linux-64::sqlite-3.30.1-hcee41ef_0
   testpath           conda-forge/noarch::testpath-0.4.4-py_0
   tk                 conda-forge/linux-64::tk-8.6.10-hed695b0_0
   traitlets          conda-forge/linux-64::traitlets-4.3.3-py38h32f6830_1
+  urllib3            conda-forge/noarch::urllib3-1.25.9-py_0
   wcwidth            conda-forge/noarch::wcwidth-0.1.9-pyh9f0ad1d_0
   webencodings       conda-forge/noarch::webencodings-0.5.1-py_1
   wheel              conda-forge/noarch::wheel-0.34.2-py_1

</pre>
</details>

Part 2:

> WIP

As discussed over on #11, the idea is to use the excellent [requests-cache](https://github.com/reclosedev/requests-cache) to implement robust caching, with the sketch of configurables presented there.

### CODE REVIEW

**Comment by bollwyvl** [comment]
Humorously, in CI, `codecov` already brings requests along, so the delta in CI is basically nil... hooray!

**Comment by blink1073** [comment]
Is the cache location selectable, or at least in a predictable place?

File: pytest_check_links/plugin.py [hunk 1]
```diff
@@ -3,19 +3,25 @@
...
-from .args import StoreExtensionsAction
+from .args import StoreExtensionsAction, StoreCacheAction
 
 _ENC = 'utf8'
 
 default_extensions = {'.md', '.rst', '.html', '.ipynb'}
 supported_extensions = {'.md', '.rst', '.html', '.ipynb'}
 
+default_cache = dict(
+    cache_name='.pytest-check-links-cache',
```
Comment (blink1073) [inline pytest_check_links/plugin.py:19]: If we `cache_name` a known path on disk we can persist that directory in CI.

Comment (blink1073) [reply pytest_check_links/plugin.py:19]: I verified by setting it with `requests_cache.install_cache('/tmp/demo_cache')`

**Comment by bollwyvl** [comment]
Yeah, as you found, you can pick where it goes. I'm still not happy with
how the config works, and haven't tested a lot of it yet, but it seems to
be mostly working.

Using the session approach avoids the requests monkeypatch, in case it's
used in a test suite that both checks links and uses requests for something
else.

Note that the default sqlite3 backend probably wouldn't work with
pytest-xdist or something... Presumably the redis/mongo backends would
handle this case. Probably with documenting!

**Comment by bollwyvl** [comment]
I've taken off WIP, with the following stomach pains:
- retries and caching are probably incompatible right now: if it got a `Retry-After`, it would... keep getting `Retry-After`. forever. seems bad.
  - [ ] there's a context manager that turns off caching
  - [ ] disable retrying if the cache is enabled
- no tests for specifying other HTTP codes than... all of them
- no tests for backend-specific args... i don't think we want to add anything fancy to the tests at present, so:
  - [ ] wait until someone breaks it
  - [ ] take out the config option, wait until someone demands it

**Comment by blink1073** [comment]
Is there not a way to remove a url from the cache?

**Comment by bollwyvl** [comment]
> Is there not a way to remove a url from the cache?

There is: it takes some [mucking about](https://github.com/reclosedev/requests-cache/issues/132#issuecomment-514923545) to properly extract the cache key. 

So notionally, it could be something like:

```py
def fetch(self, url, retries=3):
  if hasattr(self.requests_session, "cache") and retries:
    key = ...  # get the key 
    self.requests_session.cache.delete(key)
  self.requests_session.get(url)
  r = ...
```

This would keep the last failure around in the cache, which is probably a good idea...

File: pytest_check_links/plugin.py [hunk 2]
```diff
@@ -173,26 +218,31 @@ def repr_failure(self, excinfo):
...
+
+        if header is None:
+            return False
+
+        if header == '1m0s':
+            sleep_time = 60
+        else:
             try:
-                sleep_time = int(obj.headers['Retry-After'])
+                sleep_time = int(header)
```
Comment (bollwyvl) [inline pytest_check_links/plugin.py:235]: eek! guess this is happens I refactor and don't write tests!

File: test/test_cache.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    def mock_get(*args, **kwargs):
+        response = _get(*args, **kwargs)
+        if len(attempts) < 5:
+            response.status_code = 502
+            response.headers['Retry-After'] = '0'
+        attempts.append([args, kwargs])
+        return response
+
+    requests_cache.CachedSession.get = mock_get
```
Comment (bollwyvl) [inline test/test_cache.py:120]: this is not pretty, but dealing with `mock` on py27, or `pytest-mock` seemed like overkill. can use one of those if preferred. But it found the bug!

**Comment by bollwyvl** [comment]
Re: tests i'm missing: it seems like codecov isn't firing? I wonder what i broke...

**Comment by bollwyvl** [comment]
Other than the codecov thing, this is ready for review!

I'll raise some other issues about things I've been thinking about while getting this out.

**Pull request merged**